File: java/sample/src/main/java/com/github/vigneshperiasami/persisto/sample/FlatBufferSample.java
package com.github.vigneshperiasami.persisto.sample;

import com.github.vigneshperiasami.persisto.client.Flowable;
import com.github.vigneshperiasami.persisto.client.Operator;
import com.github.vigneshperiasami.persisto.client.Persisto;
import com.github.vigneshperiasami.persisto.client.ROperator;
import com.github.vigneshperiasami.persisto.client.Subject;
import com.github.vigneshperiasami.persisto.client.Subscriber;
import com.google.flatbuffers.FlatBufferBuilder;
import persisto.proto.message.Message;

import java.nio.ByteBuffer;

import static com.github.vigneshperiasami.persisto.client.Subscriber.create;

public class FlatBufferSample {

  public static void main(String[] args) throws Exception {
    Persisto persisto = Persisto.connect("127.0.0.1", 5000);

    Subject<String> messageWriter = persisto.writeSubjectByte().lift(messageWriter());
    Flowable<String> messageReader = persisto.readFlowableFixedLength().lift(messageReader());
    messageWriter.push("Hello jerry");

    messageReader.listen(create(System.out::println));
  }

  public static Operator<String, ByteBuffer> messageReader() {
    return stringSubscriber -> new Subscriber<ByteBuffer>() {
      @Override
      public void onNext(ByteBuffer data) {
        stringSubscriber.onNext(Message.getRootAsMessage(data).content());
      }

      @Override
      public void onError(Throwable err) {
        err.printStackTrace();
      }
    };
  }

  public static ROperator<String, byte[]> messageWriter() {
    return FlatBufferSample::flatMessage;
  }

  public static byte[] flatMessage(String message) {
    FlatBufferBuilder builder = new FlatBufferBuilder(1024);
    int content = builder.createString(message);
    Message.startMessage(builder);
    Message.addContent(builder, content);
    Message.addReceiver(builder, 2);
    Message.addSender(builder, 1);
    int tomMessagePayload = Message.endMessage(builder);

    builder.finish(tomMessagePayload);

    return builder.sizedByteArray();
  }
}

File: node/sample/flatbuffersample.js
const flatbuffers = require("../flatbuffers/flatbuffers").flatbuffers;
const proto = require("../sample-proto/message_generated").persisto.proto.message;
const Persisto = require("../core");

const decode = (byteArray) => {
  const buf = new flatbuffers.ByteBuffer(byteArray);
  return proto.Message.getRootAsMessage(buf);
};

const decodeOperatorFunc = (onNext) => {
  return (message) => {
    onNext(decode(message));
  };
};

const encode = (message) => {
  const builder = new flatbuffers.Builder(1024);
  const content = builder.createString(message);

  proto.Message.startMessage(builder);
  proto.Message.addContent(builder, content);
  proto.Message.addReceiver(builder, 1);
  proto.Message.addSender(builder, 2);
  const offset = proto.Message.endMessage(builder);
  builder.finish(offset);
  return builder.asUint8Array();
};

const encodeOperatorFunc = (message) => {
  message = encode(message);
  const payloadLength = Buffer.allocUnsafe(4);
  payloadLength.writeInt32BE(message.length, 0);
  return Buffer.concat([payloadLength, Buffer.from(message)]);
};

const onConnection = (persisto) => {
  const reader = persisto.readFlowableBuffer().lift(decodeOperatorFunc);
  reader.listen((message) => console.log(message.content()));

  const writer = persisto.writeSubject().lift(encodeOperatorFunc);
  writer.push("Connected to persisto!!!");
  writer.push("Hello tom!!");
  writer.push("Bye for now!!!");
};

Persisto.connect(5000, onConnection);
console.log("Listening on 5000");
